java/core/src/java/org/apache/orc/impl/SchemaEvolution.java

```diff
@@ -18,25 +18,33 @@
 
 package org.apache.orc.impl;
 
-import java.io.IOException;
 import java.util.ArrayList;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
+import java.util.regex.Pattern;
 
 import org.apache.orc.TypeDescription;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 /**
- * Take the file types and the (optional) configuration column names/types and see if there
- * has been schema evolution.
+ * Infer and track the evolution between the schema as stored in the file and the schema that has been requested by the
+ * reader.
  */
 public class SchemaEvolution {
+
+  public static class IllegalEvolutionException extends RuntimeException {
+    public IllegalEvolutionException(String msg) {
+      super(msg);
+    }
+  }
+
   private final Map<Integer, TypeDescription> readerToFile;
   private final boolean[] included;
   private final TypeDescription readerSchema;
   private static final Logger LOG = LoggerFactory.getLogger(SchemaEvolution.class);
+  private static final Pattern missingMetadataPattern = Pattern.compile("_col\\d+");
```

Comment:
You should check for WriterVersion < HIVE_4243 before bothering to check the column names. The vast majority of files from before HIVE_4243 have the synthetic column names.